### PR TITLE
Training Pipeline for ControlNet + SDXL

### PR DESCRIPTION
**Pull request by zhtmike**
- [x] 32p training result on Fill50k dataset ok
- [x] Loss converges successfully  and the canny edge of the control image can be sucessfully learned.

### CODE REVIEW

File: examples/stable_diffusion_xl/controlnet.md [hunk 1]
```diff
@@ -96,13 +104,122 @@ You can check all arguments description by running `python demo/sampling_without
...
+{"source": "source/3.png", "target": "target/3.png", "prompt": "cornflower blue circle with light golden rod yellow background"}
+```
+
+> Note: if you want to use your own dataset for training, please follow the directory and file structure shown above.
+
+### Launch training
+
+Please refer to [`scripts/run_train_base_controlnet.sh`](scripts/run_train_base_controlnet.sh).
+
+```shell
```
Comment (zhtmike) [inline examples/stable_diffusion_xl/controlnet.md:163]: `nohup` and  `> train.log 2>&1 &` is not nessesary. Users may not use or install nohup

File: examples/stable_diffusion_xl/controlnet.md [hunk 2]
```diff
@@ -96,13 +104,122 @@ You can check all arguments description by running `python demo/sampling_without
...
+    --per_batch_size 2 \
+    --group_lr_scaler 10.0 \
+    --save_ckpt_interval 10000 \
+    --max_num_ckpt 5 \
+    > train.log 2>&1 &
+```
+
+⚠️ Some key points about ControlNet + SDXL training:
+- The parameters of `zero_conv`, `input_hint_block` and `middle_block_out` blocks are randomly initialized in ControlNet, which are very hard to train. We scale up (x10 by default) the base learning rate for training parameters specifically. You can set the scale value by `args.group_lr_scaler`.
+- As mentioned in ControlNet paper[1] and [repo](https://github.com/lllyasviel/ControlNet/blob/main/docs/train.md#more-consideration-sudden-converge-phenomenon-and-gradient-accumulation), there is a sudden convergence phenomenon in ControlNet training, which means the training steps should be large enough to let the training converge SUDDENLY and then generate images following the control signals. For ControlNet + SDXL, the training steps should be even much more larger.
```
Comment (zhtmike) [inline examples/stable_diffusion_xl/controlnet.md:178]: not sure if sudden convergence happens here or just because the network is much larger, thus takes more steps to train

File: examples/stable_diffusion_xl/controlnet.md (hunk 1 above)
Comment (HaoyangLee) [reply examples/stable_diffusion_xl/controlnet.md:179]: Modified

**Review by HaoyangLee — APPROVED** [review]

File: examples/stable_diffusion_xl/gm/data/dataset.py [hunk 3]
```diff
@@ -486,6 +487,171 @@ def filter_small_image(all_images, all_captions, image_filter_size):
...
+        self.local_control_images = all_control_images
+        self.local_captions = all_captions
+
+        self.transforms = []
+        if transforms:
+            for i, trans_config in enumerate(transforms):
+                # Mapper
+                trans = instantiate_from_config(trans_config)
+                self.transforms.append(trans)
+                print(f"Adding mapper {trans.__class__.__name__} as transform #{i} " f"to the datapipeline")
```
Comment (congw729) [inline examples/stable_diffusion_xl/gm/data/dataset.py:545]: Maybe consider using `logging` instead of `print`.

File: examples/stable_diffusion_xl/gm/data/dataset.py [hunk 4]
```diff
@@ -486,6 +487,171 @@ def filter_small_image(all_images, all_captions, image_filter_size):
...
+
+        self.batched_transforms = []
+        if batched_transforms:
+            for i, bs_trans_config in enumerate(batched_transforms):
+                # Mapper
+                bs_trans = instantiate_from_config(bs_trans_config)
+                self.batched_transforms.append(bs_trans)
+                print(
```
Comment (congw729) [inline examples/stable_diffusion_xl/gm/data/dataset.py:553]: Same to the rest of all.

**Comment by zhtmike** [comment]
continue on #405
